bluetooth: host: consistent notification of subscription termination

sam-golioth opened the pull request:
Ensure that the subscription params struct is not used after releasing it back to the application.

When subscribing to a characteristic, the `struct bt_gatt_subscribe_params` pointer [must remain valid](https://github.com/zephyrproject-rtos/zephyr/blob/7a65336edc345a81508303ae97caf8fa176a5a9f/include/zephyr/bluetooth/gatt.h#L2284-L2286) until the subscription is terminated. Applications are [notified of subscription termination](https://github.com/zephyrproject-rtos/zephyr/blob/7a65336edc345a81508303ae97caf8fa176a5a9f/include/zephyr/bluetooth/gatt.h#L2157) via a call to the `params->notify()` callback with `data == NULL`. However, when the subscription is terminated via an explicit write to the GATT server's CCC descriptor, the `params` pointer is [dereferenced after the call](https://github.com/zephyrproject-rtos/zephyr/blob/7a65336edc345a81508303ae97caf8fa176a5a9f/subsys/bluetooth/host/gatt.c#L5401-L5410) to `params->notify()`, in order to call `params->subscribe()`. This makes it difficult for applications to know when they can reuse or free the `params` struct and violates the API contract as defined in the function comments.

This commit moves the call to `params->notify()` after the call to `params->subscribe()`, which ensures that calling `params->notify()` with `data == NULL` is the last use of the `params` pointer in all cases.

The `params->subscribe()` callback is only called after an explicit write to the GATT server's CCC descriptor and is not called when the subscription is terminated for other reasons (e.g. if there are multiple subscribers to the same characteristic or after a disconnect), so this callback cannot be used to know when the `params` struct has been released.

Fixes: #111427

sam-golioth committed:
bluetooth: host: consistent notification of subscription termination

Ensure that the subscription params struct is not used after releasing
it back to the application.

Signed-off-by: Sam Friedman <sam@golioth.io>

Thalley commented:
> Looks like it's failing an LE Audio BSIM test. I don't have time right now to look into this, but can provide help in a few days if you need to debugging that

Not 100% sure yet what causes this, but the PR does change the order of the `subscribe` and `notify` callbacks, and somehow this changes something so that when the failing MCS test is doing discovery (with the final step being a call to `bt_gatt_subscribe` (thus triggering a `->subscribe` callback). In the `subscribe` callback the application is setting a flag to let the application's main thread restart discovery, but where the call to `bt_gatt_discover` returns with `-ENOMEM` (with `CONFIG_BT_ATT_TX_COUNT=12`!). 

The code in the MCC (Media Control Client) is not optimal, but I'm worried that this change, with its reorder of events, may affect additional applications. Arguably since we haven't documented any specific order (I don't think at least?), then this may break more applications. 

Please let me know if I need to look further into why this is failing, and possibly provide a workaround / proper fix, or if we need to discuss this change a bit more

sam-golioth commented:
I've rebased this branch and also made changes to address the issue found by Copilot. I still need to address the failing BabbleSim tests.

sam-golioth committed:
bluetooth: mcc: check value instead of handle while unsubscribing

Checking `params->value` instead of `params->value_handle` is
a more straightforward and less brittle way to determine if a
ccc write response callback is performed due to a subscription
or unsubscription.

Signed-off-by: Sam Friedman <sam@golioth.io>

Thalley reviewed (approved):
LGTM

Thalley commented:
> Hmm, I wonder why this didn't get merge? :o

Actually I think it was due to SonarCloud blocking it due to 

> [E Maintainability Rating on New Code](https://sonarcloud.io/dashboard?id=zephyrproject-rtos_zephyr&pullRequest=99920) (required ≥ C)

@sam-golioth Unfortunately it looks like the report from SonarCloud was removed due to its age; can you try to rebase and retrigger?

jhedberg commented:
> > Hmm, I wonder why this didn't get merge? :o
> 
> Actually I think it was due to SonarCloud blocking it due to
> 
> > [E Maintainability Rating on New Code](https://sonarcloud.io/dashboard?id=zephyrproject-rtos_zephyr&pullRequest=99920) (required ≥ C)
> 
> @sam-golioth Unfortunately it looks like the report from SonarCloud was removed due to its age; can you try to rebase and retrigger?

That CI check isn't a required one, so it shouldn't be blocking the merge. Something else must be going on.

jhedberg commented:
That said:
> This branch is [2 commits ahead of](https://github.com/sam-golioth/zephyr/compare/zephyrproject-rtos%3Azephyr%3Amain...consistent_subscription_termination) and [8058 commits behind](https://github.com/sam-golioth/zephyr/compare/consistent_subscription_termination...zephyrproject-rtos%3Azephyr%3Amain) zephyrproject-rtos/zephyr:main

Simply because of that I think it'd be best to rebase. @sam-golioth could you do that? Thanks.
